src/thor/alternates.cc

```diff
@@ -56,7 +62,74 @@ void filter_alternates_by_stretch(std::vector<CandidateConnection>& connections)
   connections.erase(new_end, connections.end());
 }
 
-// Limited Sharing. Compare duration of edge segments shared between optimal path and
+// get a cost of a path between indexes 'first' and 'last'
+inline sif::Cost get_segment_cost(const std::vector<PathInfo>& path, size_t first, size_t last) {
+  auto cost = path[last].elapsed_cost - path[first].transition_cost;
+  if (first > 0)
+    cost -= path[first - 1].elapsed_cost;
+  return cost;
+};
+
+// Find a different segment for two routes. By design bidirectional astar returns routes that have
+// only one different segment. Also the first and last edges are the same for each two routes.
+std::pair<std::pair<size_t, size_t>, std::pair<size_t, size_t>>
+find_diff_segment(const std::vector<PathInfo>& path_1, const std::vector<PathInfo>& path_2) {
+
+  size_t idx_1_first = 0;
+  size_t idx_2_first = 0;
+
+  // find first different edge
+  while (idx_1_first < path_1.size() && idx_2_first < path_2.size() &&
+         path_1[idx_1_first].edgeid == path_2[idx_2_first].edgeid) {
+    ++idx_1_first;
+    ++idx_2_first;
+  }
+  // check corner cases: stop if we didn't find a different edge
+  if (idx_1_first == path_1.size())
+    return {{idx_1_first, idx_1_first}, {idx_2_first, std::max(idx_2_first, path_2.size() - 1)}};
+  else if (idx_2_first == path_2.size())
+    return {{idx_1_first, std::max(idx_1_first, path_1.size() - 1)}, {idx_2_first, idx_2_first}};
+
+  size_t idx_1_last = path_1.size() - 1;
+  size_t idx_2_last = path_2.size() - 1;
+  // find last different edge
+  while (idx_1_last > idx_1_first && idx_2_last > idx_2_first &&
+         path_1[idx_1_last].edgeid == path_2[idx_2_last].edgeid) {
+    --idx_1_last;
+    --idx_2_last;
+  }
+
+  return {{idx_1_first, idx_1_last}, {idx_2_first, idx_2_last}};
+}
+
+// Check if the candidate path contains unreasonable long detours comparing to the optimal path.
+bool validate_alternate_by_stretch(const std::vector<PathInfo>& optimal_path,
+                                   const std::vector<PathInfo>& candidate_path) {
+  const auto diff_segment = find_diff_segment(optimal_path, candidate_path);
+  const auto& idxs_opt = diff_segment.first;
+  const auto& idxs_alt = diff_segment.second;
+
+  if (idxs_opt.first == optimal_path.size()) {
```

Comment:
we could use a comment here i think, its hard to understand that this is really the second clause of the comment in the next `if` block. basically its saying, there was no start to a unique part in the optimal path, as all its edges were found in the alternate path. then the second `if` is really saying there was a unique part in the alternate path past the end of the optimal, which is just wasteful so reject it

Reply:
by the way is there any way to make that happen in gurka? im guessing its pretty difficult. you have to have two loki::search candidates on both locations so that the search paths can go the "wrong" way and then meet after the "right" way finds the optimal route

Reply:
I've met such corner case when tested this PR; but it's difficult to reproduce

Reply:
no worries, just wondered if you tried